File: WePeiYang/lostfound/src/main/java/com/twtstudio/tjwhm/lostfound/waterfall/WaterfallPresenterImpl.java
package com.twtstudio.tjwhm.lostfound.waterfall;

import com.twt.wepeiyang.commons.network.RetrofitProvider;
import com.twt.wepeiyang.commons.network.RxErrorHandler;

import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by tjwhm on 2017/7/5.
 **/

public class WaterfallPresenterImpl implements WaterfallContract.WaterfallPresenter {
    WaterfallContract.WaterfallView waterfallView;
    WaterfallApi waterfallApi;

    public WaterfallPresenterImpl(WaterfallContract.WaterfallView waterfallView) {
        this.waterfallView = waterfallView;
    }

    @Override
    public void setWaterfallData(WaterfallBean waterfallBean) {
        waterfallView.setWaterfallData(waterfallBean);
    }

    @Override
    public void loadWaterfallData(String lostOrFound, int page) {
        waterfallApi = RetrofitProvider.getRetrofit().create(WaterfallApi.class);
        waterfallApi.loadWaterData(lostOrFound, String.valueOf(page))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::setWaterfallData, new RxErrorHandler());
    }

    @Override
    public void loadWaterfallDataWithType(String lostOrFound, int page, int type) {
        if (type == -1) {
            loadWaterfallData(lostOrFound, page);
        } else {
            waterfallApi = RetrofitProvider.getRetrofit().create(WaterfallApi.class);
            waterfallApi.loadWaterfallDataWithType(lostOrFound, String.valueOf(page), String.valueOf(type))
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::setWaterfallData, new RxErrorHandler());
        }
    }


}

File: WePeiYang/lostfound2/src/main/java/com/example/lostfond2/waterfall/WaterfallPresenterImpl.kt
package com.example.lostfond2.waterfall

import com.example.lostfond2.service.LostFoundService
import com.example.lostfond2.service.MyListDataOrSearchBean
import com.twt.wepeiyang.commons.experimental.extensions.QuietCoroutineExceptionHandler
import com.twt.wepeiyang.commons.experimental.extensions.awaitAndHandle
import com.twt.wepeiyang.commons.network.RetrofitProvider
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.launch

class WaterfallPresenterImpl(var waterfallView: WaterfallContract.WaterfallView)
    : WaterfallContract.WaterfallPresenter {


    override fun setWaterfallData(waterfallBean: List<MyListDataOrSearchBean>) {
        waterfallView.setWaterfallData(waterfallBean)
    }

    override fun loadWaterfallData(lostOrFound: String, page: Int) {


//        async(UI + QuietCoroutineExceptionHandler) {
//            val dataList = when (lostOrFound) {
//                "lost" -> lostFoundService.getLost(page, 0).awaitAndHandle { it.printStackTrace() }?.data!!
//                else -> lostFoundService.getFound(page, 0).awaitAndHandle { it.printStackTrace() }?.data!!
//            }
//
//            setWaterfallData(dataList)
//        }

        launch(UI + QuietCoroutineExceptionHandler) {
            val dataList = when (lostOrFound) {
                "lost" -> LostFoundService.getLost(page, 0).await()
                else -> LostFoundService.getFound(page, 0).await()
            }

            if (dataList.error_code == -1) {
                setWaterfallData(dataList.data!!)
            }
        }
    }

    override fun loadWaterfallDataWithType(lostOrFound: String, page: Int, type: Int) {
        if (type == -1) {
            loadWaterfallData(lostOrFound, page)
        } else {
//            async(UI + QuietCoroutineExceptionHandler) {
//                val dataList = when (lostOrFound) {
//                    "lost" -> lostFoundService.getLost(page, type).awaitAndHandle { it.printStackTrace() }?.data!!
//                    else -> lostFoundService.getFound(page, type).awaitAndHandle { it.printStackTrace() }?.data!!
//                }
//
//                setWaterfallData(dataList)
//            }

            launch(UI + QuietCoroutineExceptionHandler) {
                val dataList = when (lostOrFound) {
                    "lost" -> LostFoundService.getLost(page, type).await()
                    else -> LostFoundService.getFound(page, type).await()
                }

                if (dataList.error_code == -1) {
                    setWaterfallData(dataList.data!!)
                }
            }
        }
    }
}